### ISSUE TITLE
Incorrect Doc

### ISSUE THREAD

**Issue by bessermt** [Kotlin/kotlinx.coroutines#1394]
Bridging blocking and non-blocking worlds
The first example mixes non-blocking delay(...) and blocking Thread.sleep(...) in the same code

Makes no sense. Thread sleep only blocks current thread. No idea what you are trying to say here.

**Comment by qwwdfsad** [comment]
We are trying to say that `Thread.sleep` blocks the current thread, while `delay` does not.
It means that a single thread can process only one task with `Thread.sleep` inside it at once, but almost unbounded amount of tasks with `delay`

**Comment by halimpuckjava** [comment]
suspend function delay() doesn't block thread because its dispatched in other thread, so the current thread where coroutine is launched it will be free to run other task, this coroutine also its suspend (unbound from any thread) execution until delay finish, when delay() finish the coroutine can continue execution either in last thread or in other thread pool

**Comment by labermt** [comment]
I'm sorry @halimpuckjava, I can't understand your English. Maybe try to correct your English and I'll understand better. Thanks.

**Comment by labermt** [comment]
I did some research and it appears the docs are wrong in numerous ways. There are so many mistakes that I suggest they are thrown away and restarted. I can't do it because I'm new to Kotlin and I don't have the time. 

Here are some fundamental problems just to give you an idea why they are not small mistakes that can be corrected. First the title:

Coroutine basics

In English, titles are caplitalized unless the term is minor. Even if corrected, the section is not "basic" in any sense of the word. Basics should start with reader prerequisites and links to a glossary of terms. Showing code that is not a typical use case is anything but "basic."

In addition, there are problems with correct use of terms. There is no description if the docs are talking about hardware or software threads. It appears, but is impossible to tell, the docs are referring to hardware threads while conflating them with how software threads work. 

The problems go deeper. The doc says "Essentially, coroutines are light-weight threads." This is factually wrong and worse, ... misleading. A coroutine is not a thread is any sense of the term and should not be compared. When you mislead a reader, you send them on a bad path that will negatively impact their ability to learn the material. 

A coroutine is a set of computer instructions managed at the language level. The language will schedule with the OS when to run the coroutine instructions as software threads become available. When a coroutine is blocked using the delay() function, the language may choose to pause the coroutine and place another coroutine on that software thread. This means that while the coroutine is blocked, the thread can continue doing useful work. In contrast, the sleep() function blocks the coroutine as well as the thread, not allowing either to do useful work. 

In short, everything in this doc from grammar to content is incorrect and needs to be thrown out. 

That said, it is possible that since I'm new, I may not be understanding the material correctly. Regardless, either I don't understand and everything I've said above is wrong, or the docs are so poorly written that I'm unable to properly understand the material. Either way, the docs are less than useless. They are damaging the reader's ability to learn the material.

**Comment by elizarov** [comment]
Thanks for feedback. These docs are used to be named "Guide to coroutines by example" and assume certain (quite high) initial level of knowledge on the part of the reader. They are indeed scheduled to be rewritten at some point in the future into a more introductory form that would be better suited for wider audience. 

However, we do perform minor corrections, typo fixes, and updates, so we do accept reports and PRs with specific issues and correction. Since this issue is not specific, I'm closing it.

**Comment by zach-klippenstein** [comment]
Writing technical documentation is hard, and I think it's possible and much more productive to give feedback _without_ using hyperbole and insulting the people who worked on it. 

I strongly disagree with "everything is wrong" and "the docs are less than useless". I found them very helpful when starting out, and the last time I read through them they were accurate, at least at a high level. You bring up some good points but everything you mention can be addressed with some edits, and doesn't require rewriting from scratch, especially minor grammar nits.

**Comment by labermt** [comment]
@zach-klippenstein ... I did not use [Ad hominem](https://en.wikipedia.org/wiki/Ad_hominem), and never insulted "the people" that worked on it. I also never used "hyperbole", and meant everything I said literally. Everything in that document is wrong and is less than useless. Every document has an intended purpose and this one missed the "basics" target. That's why everything in the document is wrong. It doesn't address the intended purpose given the title and placement of the document at: 

https://kotlinlang.org/docs/reference/coroutines/basics.html. 

You missed the point that I don't have the time to go though every mistake in the document. There are too many mistakes for me to do that. Even if I did, as stated by elizarov, the doc has an unstated prerequisite of a high level of knowledge but it is the first document on the topic and is listed as "basic", thereby leading readers to go there first when starting out. It is simply wrong at a fundamental level to mark a document as basic and expect advanced knowledge. That can't be fixed with edits. It needs to be removed as the starting "basic" document. If you don't do that, those looking for introductory information on coroutines will be in a worse position than if they didn't read the document at all, "less than useless". 

Sure minor grammar nits are fixable, but minor grammar problems show a lack of attention to detail that makes me think the writer considered this a draft, meant to be either thrown away or completely reorganized. It also makes me wonder if the document was ever sent out for review to the intended "beginner" audience because I would suspect these grammar problems would have been caught if reviewed.  

What I do know is that the parts I checked are technically incorrect and you are incorrect they are "accurate." My research says a coroutine is in no way a light-weight hardware thread nor a light-weight software thread. Yet, the very first statement about coroutines is: "Essentially, coroutines are light-weight threads." Maybe you don't understand the meaning of the word essentially? This is incorrect and you should simply admit it or prove me wrong. By acting insulted it appears you are personally worried that you don't understand the content of what was written in the document. 

If you want to take criticism of work as a personal insult, that's your right as a millennial. Yes, now you've been insulted. See the difference? When I call you a millennial, I'm reading what you wrote and making assumptions about your age and being raised in a culture where any criticism is seen though a personal lens, the stereotypical millennial behavior. I don't know if you are young or old, but your age is a personal attribute and has nothing to do with the content of the material. If I had done that, I would deserve your comments, but I didn't. Go back and read what I wrote. There is not a single _Ad hominem_ attack. 

If you wrote this document, I'm sorry to tell you the material is not close to acceptable for it's intended purpose and can't be addressed with some edits. It has everything from minor grammar problems to having fundamentally wrong information regarding threading. Even if all that is corrected, it is still not "basic" and is therefore not fixable with edits.

I'm sorry you took this as an insult, but that's on you and not what I wrote. Please go back and read what I wrote and not don't dismiss the facts because you felt insulted. Good luck.
